overlord/healthstate/healthstate_test.go

```diff
@@ -0,0 +1,194 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package healthstate_test
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord"
+	"github.com/snapcore/snapd/overlord/healthstate"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaptest"
+	"github.com/snapcore/snapd/store/storetest"
+	"github.com/snapcore/snapd/testutil"
+)
+
+func TestHealthState(t *testing.T) { check.TestingT(t) }
+
+type healthSuite struct {
+	testutil.BaseTest
+	o       *overlord.Overlord
+	se      *overlord.StateEngine
+	state   *state.State
+	hookMgr *hookstate.HookManager
+	info    *snap.Info
+}
+
+var _ = check.Suite(&healthSuite{})
+
+func (s *healthSuite) SetUpTest(c *check.C) {
+	s.BaseTest.SetUpTest(c)
+	s.AddCleanup(healthstate.MockCheckTimeout(time.Second))
+	dirs.SetRootDir(c.MkDir())
+
+	s.o = overlord.Mock()
+	s.state = s.o.State()
+
+	var err error
+	s.hookMgr, err = hookstate.Manager(s.state, s.o.TaskRunner())
+	c.Assert(err, check.IsNil)
+	s.se = s.o.StateEngine()
+	s.o.AddManager(s.hookMgr)
+	s.o.AddManager(s.o.TaskRunner())
+
+	healthstate.Init(s.hookMgr)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	snapstate.ReplaceStore(s.state, storetest.Store{})
+	sideInfo := &snap.SideInfo{RealName: "test-snap", Revision: snap.R(42)}
+	snapstate.Set(s.state, "test-snap", &snapstate.SnapState{
+		Sequence: []*snap.SideInfo{sideInfo},
+		Current:  snap.R(42),
+		Active:   true,
+		SnapType: "app",
+	})
+	s.info = snaptest.MockSnapCurrent(c, "{name: test-snap, version: v1}", sideInfo)
+}
+
+func (s *healthSuite) TearDownTest(c *check.C) {
+	s.hookMgr.StopHooks()
+	s.se.Stop()
+	s.BaseTest.TearDownTest(c)
+}
+
+func (s *healthSuite) TestHealthNoHook(c *check.C) {
+	s.testHealth(c, false, false)
```

Comment:
I'm not a fan of these booleans as I need to look at how testHealth is defined to see what they do. Granted its super close by and has good names so I'm probably overly pedantic. I looked at this a bit and the only other suggestion I have is something like this:
```
--- a/overlord/healthstate/healthstate_test.go
+++ b/overlord/healthstate/healthstate_test.go
@@ -90,26 +90,37 @@ func (s *healthSuite) TearDownTest(c *check.C) {
 	s.BaseTest.TearDownTest(c)
 }
 
+type healthHookTestCondition int
+
+const (
+	noHook = iota
+	badHook
+	goodHook
+	captainHook
+)
+
 func (s *healthSuite) TestHealthNoHook(c *check.C) {
-	s.testHealth(c, false, false)
+	s.testHealth(c, noHook)
 }
 
 func (s *healthSuite) TestHealthFailingHook(c *check.C) {
-	s.testHealth(c, true, true)
+	s.testHealth(c, badHook)
 }
 
 func (s *healthSuite) TestHealth(c *check.C) {
-	s.testHealth(c, true, false)
+	s.testHealth(c, goodHook)
 }
 
-func (s *healthSuite) testHealth(c *check.C, withScript, failScript bool) {
+func (s *healthSuite) testHealth(c *check.C, cond healthHookTestCondition) {
 	var cmd *testutil.MockCmd
-	if failScript {
+	switch cond {
+	case badHook:
 		cmd = testutil.MockCommand(c, "snap", "exit 1")
-	} else {
+	default:
 		cmd = testutil.MockCommand(c, "snap", "exit 0")
 	}
-	if withScript {
+
+	if cond != noHook {
 		hookFn := filepath.Join(s.info.MountDir(), "meta", "hooks", "check-health")
 		c.Assert(os.MkdirAll(filepath.Dir(hookFn), 0755), check.IsNil)
 		// the hook won't actually be called, but needs to exist
@@ -150,19 +161,20 @@ func (s *healthSuite) testHealth(c *check.C, withScript, failScript bool) {
 	err = s.state.Get("health", &healths)
 	s.state.Unlock()
 
-	if failScript {
+	switch cond {
+	case badHook:
 		c.Assert(status, check.Equals, state.ErrorStatus)
-	} else {
+	default:
 		c.Assert(status, check.Equals, state.DoneStatus)
 	}
-	if withScript {
+	if cond != noHook {
 		c.Assert(err, check.IsNil)
 		c.Assert(healths, check.HasLen, 1)
 		c.Assert(healths["test-snap"], check.NotNil)
 		health := healths["test-snap"]
 		c.Check(health.Revision, check.Equals, snap.R(42))
 		c.Check(health.Status, check.Equals, healthstate.UnknownStatus)
-		if failScript {
+		if cond == badHook {
 			c.Check(health.Message, check.Equals, "hook failed")
 			c.Check(health.Code, check.Equals, "snapd-hook-failed")
 		} else {
```
But feel free to ignore.